presto-main/src/main/java/com/facebook/presto/metadata/SessionFunctionRegistry.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.metadata;
+
+import com.facebook.presto.common.CatalogSchemaName;
+import com.facebook.presto.common.function.QualifiedFunctionName;
+import com.facebook.presto.spi.function.BasicSignature;
+import com.facebook.presto.spi.function.FunctionHandle;
+import com.facebook.presto.spi.function.FunctionImplementationType;
+import com.facebook.presto.spi.function.FunctionMetadata;
+import com.facebook.presto.spi.function.FunctionNamespaceTransactionHandle;
+import com.facebook.presto.spi.function.FunctionRegistry;
+import com.facebook.presto.spi.function.Parameter;
+import com.facebook.presto.spi.function.ScalarFunctionImplementation;
+import com.facebook.presto.spi.function.Signature;
+import com.facebook.presto.spi.function.SqlFunction;
+import com.facebook.presto.spi.function.SqlInvokedFunction;
+import com.facebook.presto.spi.function.SqlInvokedScalarFunctionImplementation;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.facebook.presto.server.SerializationUtils.BasicSignatureSerDe;
+import static com.facebook.presto.server.SerializationUtils.SqlInvokedFunctionSerDe;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.stream.Collectors.collectingAndThen;
+import static java.util.stream.Collectors.toMap;
+
+public class SessionFunctionRegistry
+        implements FunctionRegistry<SqlFunction>
+{
+    public static final CatalogSchemaName SESSION_NAMESPACE = new CatalogSchemaName("presto", "session");
+
+    private final ImmutableMap<BasicSignature, SqlInvokedFunction> sessionFunctions;
+
+    public SessionFunctionRegistry(Map<String, String> sessionFunctions)
+    {
+        this.sessionFunctions = sessionFunctions.entrySet().stream().collect(collectingAndThen(toMap(
+                e -> BasicSignatureSerDe.deserialize(e.getKey()),
+                e -> SqlInvokedFunctionSerDe.deserialize(e.getValue())), ImmutableMap::copyOf));
+    }
+
+    @Override
+    public Collection<SqlFunction> listFunctions()
+    {
+        return new ArrayList<>(sessionFunctions.values());
```

Comment:
Does it has to be array list, since `sessionFunctions.values()` is already a Collection?

Reply:
Yes, we need to create a new List because sessionFunctions.values() is Collection<SqlInvokedFunction> and we need Collection<SqlFunction>